### PR TITLE
Apt 246 substitute second lender nft to datum

### PR DESCRIPTION
**Pull request by tomazvila**
The second copy of Lender NFT that is sent to Collateral.hs and then to Interest.hs can be substituted by using a datum. This would remove a lot of complicated checks

### CODE REVIEW

**Review by lukipuki — APPROVED** [review]

File: src/Collateral.hs [hunk 1]
```diff
@@ -100,31 +101,28 @@ mkValidator contractInfo@ContractInfo{..} dat rdm ctx = validate
...
-
-    filterValues :: Value -> [(CurrencySymbol, TokenName, Integer)]
-    filterValues v = filter lenderNftFilter $ flattenValue v
+    validateBorrowerNftBurn :: Bool
+    validateBorrowerNftBurn = any (\(cs, tn, n) -> cs == borrowersNFT dat && tn == borrower && n == (-1)) (U.mintFlattened ctx)
 
-    containsFlattenedValue :: [(CurrencySymbol, TokenName, Integer)] -> (CurrencySymbol, TokenName, Integer) -> Bool
-    containsFlattenedValue xs (cs, tn, _) = foldr (\(cs', tn', _) acc -> (cs' == cs && tn' == tn) || acc) False xs
+    getCollateralScHashes :: [DatumHash]
+    getCollateralScHashes = map fst (scriptOutputsAt interestscvh (U.info ctx))
```
Comment (mikx98) [inline src/Collateral.hs:108]: I think the naming is misleading. It actually returns the datum hashes located on the address of the Interest script.
```suggestion
    getInterestScHashes :: [DatumHash]
    getInterestScHashes = map fst (scriptOutputsAt interestscvh (U.info ctx))
```

File: src/Interest.hs [hunk 2]
```diff
@@ -47,33 +48,49 @@ data ContractInfo = ContractInfo
     { lenderNftCs  :: !CurrencySymbol
     } deriving (Show, Generic, ToJSON, FromJSON)
 
+data InterestDatum = InterestDatum
+    { borrowersNFT          :: !CurrencySymbol
+    , borrowersPkh          :: !PaymentPubKeyHash
```
[the remark this answers is not in the document]
Comment (tomazvila) [reply src/Interest.hs]: removed

File: src/Interest.hs [hunk 3]
```diff
@@ -47,33 +48,49 @@ data ContractInfo = ContractInfo
...
+    , interestcs            :: !CurrencySymbol
+    , interestamnt          :: !Integer
+    , collateralcs          :: !CurrencySymbol
+    , repayinterval         :: !POSIXTime
+    , liquidateNft          :: !CurrencySymbol
+    , collateraltn          :: !TokenName -- collateral token name
+    , collateralamnt        :: !Integer   -- amount of collateral
+    , collateralFactor      :: !Integer   -- Colalteral factor used for liquidation
+    , liquidationCommission :: !Integer   -- How much % borrower will pay for lender when liquidated (before time passes)
+    , requestExpiration     :: !POSIXTime
```
[the remark this answers is not in the document]
Comment (tomazvila) [reply src/Interest.hs]: removed

File: src/Interest.hs [hunk 4]
```diff
@@ -47,33 +48,49 @@ data ContractInfo = ContractInfo
...
+    , loantn                :: !TokenName
+    , loancs                :: !CurrencySymbol
+    , loanamnt              :: !Integer
+    , interesttn            :: !TokenName
+    , interestcs            :: !CurrencySymbol
+    , interestamnt          :: !Integer
+    , collateralcs          :: !CurrencySymbol
+    , repayinterval         :: !POSIXTime
+    , liquidateNft          :: !CurrencySymbol
+    , collateraltn          :: !TokenName -- collateral token name
```
[the remark this answers is not in the document]
Comment (tomazvila) [reply src/Interest.hs]: removed

**Review by mikx98 — APPROVED** [review]

File: src/Collateral.hs [hunk 5]
```diff
@@ -143,13 +149,8 @@ mkValidator contractInfo@ContractInfo{..} dat rdm ctx = validate
...
     checkMintTnName = traceIfFalse "invalid time token name" (maybe False tokenNameIsCorrect getTimeTokenName)
 
-    inputHasBurntLNft :: CurrencySymbol -> TokenName -> Bool
-    inputHasBurntLNft cs tn = case U.ownValue ctx of
-      Just v  -> valueOf v cs tn == 1
-      Nothing -> False
-
     checkLNftsAreBurnt :: Bool
-    checkLNftsAreBurnt = traceIfFalse "2 Lender Nfts not burnt" (any (\(cs, tn, n) -> inputHasBurntLNft cs tn && n == (-2)) (U.mintFlattened ctx))
+    checkLNftsAreBurnt = traceIfFalse "Lender Nft not burnt" (any (\(cs, tn, n) -> cs == lenderNftCs && tn == lenderNftTn dat && n == (-1)) (U.mintFlattened ctx))
```
Comment (lukipuki) [inline src/Collateral.hs:153]: Here you know cs&tn, you can use `valueOf` or `assetClassValueOf`.

**Review by mikx98 — APPROVED** [review]

**Review by mikx98 — APPROVED** [review]
Time related logic seems ok now

**Pull request merged**